### DISCUSSION TITLE
Other DACs support

### DISCUSSION THREAD

**Discussion by matemaciek** [probonopd/MiniDexed#124] [Q&A] [answered]
Hi!

I'm trying to build device with [raspiaudio Ultra+](https://raspiaudio.com/produit/ultra), but id does not output any sound. I tested with [regular raspiaudio](https://raspiaudio.com/produit/audio) and with that everyhing works, so my guess is that DAC in Ultra+ (WM8960) is not supported by circle? Or maybe I am missing something?

**Comment by probonopd** [comment, marked as answer]
Welcome @matemaciek.
Correct. It is not supported yet. I'd like to see it supported, too.

Maybe you can help to get it supported?
Please see https://github.com/dwhinham/mt32-pi/issues/273 for the details.

**Reply by matemaciek** [reply]
I'll see what I can do (-:

**Reply by matemaciek** [reply]
How is MiniDexed connected to mt32? I don't see it referenced anywhere in the repo.

**Reply by probonopd** [reply]
They both use [Circle](https://github.com/rsta2/circle/), the environment that allows applications to run directly on Raspberry Pi hardware without an operating system such as Linux. If this gets implemented for mt32, it should be easy to port it over to other Circle-using projects such as MiniDexed.

**Reply by krunchr** [reply]
Are there any chances for support of class-compliant USB audio interfaces in the future?

**Reply by matemaciek** [reply]
I see that Circle [included mt32 code for inicialization of i2c DACs](https://github.com/rsta2/circle/blob/3141e5237f275c08828051f96b34a09fc256c2de/lib/i2ssoundbasedevice.cpp#L426), so it looks like Circle would be the best place to add support for other DACs.

**Reply by probonopd** [reply]
@matemaciek: Yes, I am sure that once _any_ Circle-using project has a working implementation, it can be upstreamed to Circle.

**Reply by probonopd** [reply]
@krunchr 

> Are there any chances for support of class-compliant USB audio interfaces in the future?

Not in the near future. But one should never say "never": https://github.com/rsta2/circle/issues/222.
If you are looking for an inexpensive "DAC" without soldering, [this hack](https://github.com/probonopd/MiniDexed/wiki/Hardware#hdmi-to-audio) might fit your needs?

**Reply by krunchr** [reply]
> Not in the near future. But one should never say "never": [rsta2/circle#222](https://github.com/rsta2/circle/issues/222).
> If you are looking for an inexpensive "DAC" without soldering, [this hack](https://github.com/probonopd/MiniDexed/wiki/Hardware#hdmi-to-audio) might fit your needs?

Thanks for your fast reply. It's not so much about saving money as it is about continuing to use an old and unused USB audio interface ;) Anyway, I just ordered a raspiaudio dac.

**Reply by probonopd** [reply]
In the meantime, you should get at least _some_ audio out of the headphone jack ;)

**Reply by matemaciek** [reply]
Any hints on debugging/development workflow? Are logs sent to UART by default?

**Reply by probonopd** [reply]
Logging: https://github.com/probonopd/MiniDexed/wiki/Development#logging

**Reply by matemaciek** [reply]
It's alive!

[I've added WM8960 support to Circle](https://github.com/matemaciek/circle/commit/cb2b3843b3efa84d9ccf4d2f1c739b95b1404ec3), let's see wether it will be accepted mainstream. No changes in MiniDexed are needed, as after my changes Circle detects which DAC is present on i2c bus.

**Reply by probonopd** [reply]
@matemaciek you are a genius!

**Reply by matemaciek** [reply]
It was a matter of perfect timing - when I started trying to reverse engineer i2c configuration, [this appeared](https://github.com/RASPIAUDIO/ULTRA/blob/main/ultra.c) (-:

**Reply by probonopd** [reply]
@matemaciek It works! It works! It works!  :+1:  Details: https://github.com/rsta2/circle/pull/286#issuecomment-1105544344
